nmsg/config_file.c

```diff
@@ -0,0 +1,387 @@
+};
+
+struct _config_file_section {
+	char 					*name;
+	size_t 					name_len;
+	ISC_LIST(struct config_file_item)	items;
+	ISC_LINK(struct _config_file_section)	link;
+};
+
+struct config_file {
+	struct _config_file_section		*root;
+	ISC_LIST(struct _config_file_section)	sections;
+};
+
+static const char *_my_strnchr(const char *data, size_t data_len, char ch);
+
+static struct config_file_item *_config_file_find_item(struct config_file_item *items, const char *key, size_t key_len);
+
+static void _config_file_set_item_value(struct _config_file_item_value *item, const char *key, size_t key_len,
+				      const char *value, size_t value_len);
+
+static struct config_file_item *_config_file_create_item(const char *key, size_t key_len,
+						     const char *value, size_t value_len);
+
+static struct _config_file_section *_config_file_add_section(struct config_file *config, bool root,
+							 const char *name, size_t name_len);
+
+static struct _config_file_section *_config_file_find_section(struct config_file *config, const char *name, size_t name_len);
+
+static void _config_file_destroy_item(struct config_file_item *item);
+
+static void _config_file_destroy_items(struct _config_file_section *section);
+
+static void _config_file_destroy_sections(struct config_file *config);
+
+/*
+ * PRIVATE FUNCTIONS
+ */
+
+static const char *
+_my_strnchr(const char *data, size_t data_len, char ch) {
+	while(data_len > 0) {
+		if (*data == ch)
+			return data;
+		--data_len;
+		++data;
+	}
+
+	return NULL;
+}
+
+static struct config_file_item *
+_config_file_find_item(struct config_file_item *items, const char *key, size_t key_len) {
+	while(items != NULL) {
+		if (strncmp(items->item->key, key, key_len) == 0)
+			return items;
+		items = ISC_LIST_NEXT(items, link);
+	}
+	return NULL;
+}
+
+static void
+_config_file_set_item_value(struct _config_file_item_value *item, const char *key, size_t key_len, const char *value, size_t value_len) {
+	my_free(item->key);
+	my_free(item->value);
+	item->key = my_strndup(key, key_len);
+	item->value = my_strndup(value, value_len);
+}
+
+static struct config_file_item *
+_config_file_create_item(const char *key, size_t key_len, const char *value, size_t value_len) {
+	struct config_file_item *item;
+	item = my_calloc(1, sizeof(struct config_file_item));
+	item->item = my_calloc(1, sizeof(struct _config_file_item_value));
+	_config_file_set_item_value(item->item, key, key_len, value, value_len);
+	return item;
+}
+
+static bool
+_config_file_add_item(struct _config_file_section *section, const char *data, size_t data_len) {
+	size_t key_len = 0;
+	size_t value_len = 0;
+	const char *divider;
+	struct config_file_item *item;
+
+	divider = _my_strnchr(data, data_len, _EQUAL_DIV);
+	if (divider == NULL)
+		return false;
+
+	key_len = divider - data;
+	++divider;
+	value_len = data_len - key_len - 1;
+
+	while((data[key_len - 1] == ' ' || data[key_len - 1] == '\t') && key_len > 1)
```

Comment:
maybe change of these to standard ctype.h isblank():  checks for a blank character; that is, a space or a tab.